File: app/src/main/java/su/linka/pictures/Set.java
package su.linka.pictures;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.util.ArrayList;
import java.util.UUID;

public class Set {
    private final SetManifest manifest;
    private final File folder;
    public Set(SetManifest manifest, File folder) {
        this.manifest = manifest;
        this.folder = folder;

    }

    public SetManifest getManifest() {
        return manifest;
    }

    public Bitmap getBitmap(String imagePath) {
        if(imagePath==null) return null;
        File file = new File(folder, imagePath);
        return Utils.readBitmapFromFile(file);
    }

    public File getAudioFile(String audioPath) {
        return new File(folder, audioPath);
    }

    public File copyAudioFile(File currentAudioFile) throws IOException {
        return Utils.copy(currentAudioFile, new File(folder, currentAudioFile.getName()));
    }


    public File saveBitmap(Bitmap bm) {
        folder.mkdir();
        return Utils.saveBitmapToFile(folder, UUID.randomUUID().toString()+".png",bm,Bitmap.CompressFormat.PNG,100);

    }

    public void addCard(int pos, Card result) {
        ArrayList<Card> cards = getManifest()
                .cards;
        int size = cards.size();
        if(pos>=size){
            while (cards.size()<pos){
                cards.add(new Card(0, 3 ));
            }
            cards.add(result);

        } else {
            cards.set(pos, result);
        }

    }

    public void writeConfig() {
        JSONObject object;
        try {
            object = manifest.toJSONObject();
        } catch (JSONException e) {
            e.printStackTrace();
            return;
        }
        File file = manifest.file;

        try {
            Writer output = null;
            output = new BufferedWriter(new FileWriter(file));
            output.write(object.toString());
            output.close();

        } catch (Exception e) {
        }
    }

    public File getFolder() {
        return folder;
    }
}

File: shared/src/androidMain/kotlin/su/linka/pictures/Set.kt
package su.linka.pictures

import android.graphics.Bitmap
import org.json.JSONException
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.util.UUID

class Set(
    private val manifest: SetManifest,
    private val folder: File
) {

    fun getManifest(): SetManifest = manifest

    fun getBitmap(imagePath: String?): Bitmap? {
        if (imagePath == null) return null
        val file = File(folder, imagePath)
        return Utils.readBitmapFromFile(file)
    }

    fun getAudioFile(audioPath: String?): File? {
        if (audioPath == null) return null
        return File(folder, audioPath)
    }

    @Throws(IOException::class)
    fun copyAudioFile(currentAudioFile: File): File {
        folder.mkdirs()
        val source = currentAudioFile.absoluteFile
        val destination = File(folder, source.name).absoluteFile
        if (destination.absolutePath == source.absolutePath) {
            return destination
        }
        val target = if (!destination.exists()) {
            destination
        } else {
            resolveUniqueDestination(source.name)
        }
        return Utils.copy(source, target)
    }

    fun saveBitmap(bitmap: Bitmap): File {
        folder.mkdirs()
        return Utils.saveBitmapToFile(
            folder,
            "${UUID.randomUUID()}.png",
            bitmap,
            Bitmap.CompressFormat.PNG,
            100
        )
    }

    fun addCard(pos: Int, card: Card) {
        val cards = manifest.cards
        if (pos >= cards.size) {
            while (cards.size < pos) {
                cards.add(Card(0, 3))
            }
            cards.add(card)
        } else {
            cards[pos] = card
        }
    }

    fun writeConfig() {
        val json = try {
            manifest.toJSONObject()
        } catch (error: JSONException) {
            error.printStackTrace()
            return
        }
        val configFile = manifest.configFile
        configFile.parentFile?.mkdirs()
        try {
            BufferedWriter(FileWriter(configFile)).use { writer ->
                writer.write(json.toString())
            }
        } catch (_: Exception) {
        }
    }

    fun getFolder(): File = folder

    private fun resolveUniqueDestination(originalName: String): File {
        val dotIndex = originalName.lastIndexOf('.')
        val baseName = if (dotIndex != -1) originalName.substring(0, dotIndex) else originalName
        val extension = if (dotIndex != -1) originalName.substring(dotIndex) else ""
        var index = 1
        while (true) {
            val candidate = File(folder, "$baseName-$index$extension").absoluteFile
            if (!candidate.exists()) {
                return candidate
            }
            index++
        }
    }
}
